core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -981,9 +981,22 @@ public void onSuccess() {
                                         }
 
                                         @Override
-                                        public void onFailure(Throwable t) {
-                                            // TODO: handle catastrophic non-channel failures
-                                            onReplicaFailure(nodeId, exp);
+                                        public void onFailure(Throwable shardFailedError) {
+                                            if (shardFailedError instanceof ShardStateAction.NoLongerPrimaryShardException) {
+                                                ShardRouting primaryShard = indexShardReference.routingEntry();
+                                                String message = String.format(Locale.ROOT, "primary shard [%s] was demoted while failing replica shard [%s] for [%s]", primaryShard, shard, exp);
+                                                // we are no longer the primary, fail ourselves and start over
+                                                indexShardReference.failShard(message, shardFailedError);
+                                                Releasables.close(indexShardReference);
+                                                try {
+                                                    channel.sendResponse(new RetryOnPrimaryException(shardId, message));
+                                                } catch (IOException channelException) {
+                                                    logger.warn("{} failed to send response back to client for action [" + transportReplicaAction + "] after failing demoted primary shard [{}]", channelException, shardId, primaryShard);
+                                                }
+                                            } else {
+                                                assert false : shardFailedError;
+                                                forceFinishAsFailed(shardFailedError);
```

Comment:
I thought we decided to keep old behavior here? This would mean onReplicaFailure ?